src/stamps.ts

```diff
@@ -1,28 +1,235 @@
+            reject(e)
+          })
+      })
+      .catch(e => {
+        logger.error('error: failed to buy postage stamp', e)
+        reject(e)
+      }),
+  )
+}
+
+export class StampsManager {
+  private stamp?: string
+  private usableStamps?: DebugPostageBatch[]
+  public readonly enabled: boolean
+  private interval?: ReturnType<typeof setInterval>
+
+  constructor(
+    {
+      POSTAGE_STAMP,
+      POSTAGE_DEPTH,
+      POSTAGE_AMOUNT,
+      POSTAGE_USAGE_THRESHOLD,
+      POSTAGE_USAGE_MAX,
+      POSTAGE_TTL_MIN,
+      POSTAGE_REFRESH_PERIOD,
+    }: Args = {},
+    BEE_DEBUG_API_URL?: string,
+  ) {
+    this.enabled = true
+
+    if (POSTAGE_STAMP) this.stamp = POSTAGE_STAMP
+    else if (POSTAGE_DEPTH && POSTAGE_AMOUNT && BEE_DEBUG_API_URL) {
+      const refreshPeriod = Number(POSTAGE_REFRESH_PERIOD || 60_000)
+      this.start(
+        Number(POSTAGE_DEPTH),
+        POSTAGE_AMOUNT,
+        Number(POSTAGE_USAGE_THRESHOLD || '0.7'),
+        Number(POSTAGE_USAGE_MAX || '0.95'),
+        Number(POSTAGE_TTL_MIN || (refreshPeriod / 1_000) * 5),
+        refreshPeriod,
+        new BeeDebug(BEE_DEBUG_API_URL),
+      )
+    } else if (POSTAGE_DEPTH || POSTAGE_AMOUNT || BEE_DEBUG_API_URL) {
+      throw new Error('Please provide POSTAGE_DEPTH, POSTAGE_AMOUNT and BEE_DEBUG_API_URL environment variable')
+    } else this.enabled = false
+  }
+
+  /**
+   * Get postage stamp that should be replaced in a the proxy request header
+   *
+   * @return Postage stamp that should be used by the proxy
+   *
+   * @throws Error if there is no postage stamp
+   */
+  get postageStamp(): string {
+    if (this.stamp) return this.stamp
+
+    if (this.usableStamps && this.usableStamps[0]) return this.usableStamps[0].batchID
+
+    throw new Error('No postage stamp')
+  }
+
+  /**
+   * Refresh stamps from the bee node and if needed buy new stamp
+   *
+   * @param depth Postage stamps depth
+   * @param amount Postage stamps amount
+   * @param usageThreshold Threshold when new postage stamp should be bought
+   * @param maxUsage Maximal usage of the stamp to be usable by the proxy
+   * @param minTTL Minimal TTL of the stamp to be usable by the proxy
+   * @param beeDebug Connection to debug endpoint for checking/buying stamps
+   */
+  public async refreshStamps(
+    depth: number,
+    amount: string,
+    usageThreshold: number,
+    maxUsage: number,
+    minTTL: number,
+    beeDebug: BeeDebug,
+  ): Promise<void> {
+    try {
+      const stamps = await beeDebug.getAllPostageBatch()
+
+      // Get all usable stamps sorted by usage from most used to least
+      this.usableStamps = filterUsableStamps(stamps, depth, amount, maxUsage, minTTL)
+
+      // Check if the least used stamps is starting to get full and if so purchase new stamp
+      const leastUsed = this.usableStamps[this.usableStamps.length - 1]
+
+      if (!leastUsed || getUsage(leastUsed) > usageThreshold) {
+        buyNewStamp(depth, amount, beeDebug).then(() => {
```

Comment:
I would use `await`. If the buying stamps fail then the error is not caught and is silenced. In this way we won't register the race conditions if they happen.